Software keyboard support.

kevinkoenig opened the pull request:
Closes #116

- Centered keyboard
- Cleaned up all the magic numbers
- provide visual feedback on TouchStart
- fixes hit boxes to be halfway before the letter and halfway after.  the old logic was at the first pixel of the letter to the right, making key selection haphazard.  
- Fixed the bitmap position of the backspace.
- fixed issue with having edited fields be off screen

most of the work was in libopenui which will have a separate pull request

pfeerick commented:
I think this will just need the `libopenui` submodule switching back to EdgeTX _after_ https://github.com/EdgeTX/libopenui/pull/14 is merged. If the github builds manage to build this, I'll try it out later today on the TX16S and NV14, otherwise it will be a tomorrow job.

pfeerick commented on radio/src/targets/horus/CMakeLists.txt:
```diff
@@ -1,7 +1,6 @@
 option(DISK_CACHE "Enable SD card disk cache" ON)
 option(UNEXPECTED_SHUTDOWN "Enable the Unexpected Shutdown screen" ON)
 option(FLYSKY_HALL_STICKS "Enable FlySky Hall Sticks" OFF)
-option(IMU_LSM6DS33 "Enable I2C2 and LSM6DS33 IMU" OFF)
```
I think you need to rebase or merge the latest state of `main` into your branch, because this and the two references to IMU_LSM6DS33 were merged 4 days ago. ;)

kevinkoenig replied:
how do i do that?  do i do a pull and then reissue the pull request?

rotorman replied:
Have a look at the following link:
https://github.com/EdgeTX/edgetx/pull/285#issuecomment-883332611

pfeerick replied:
With a rebase - you will need to do a forced push after getting it up to date again. With a merge, it will add another commit, and will push normally, and update the PR to suit (since the PR is a reflection of the state of the branch the PR is based on). rebase is cleaner for commit history, but can be a pain if it is the first time you've done on.

pfeerick commented:
First impression... this is really good! :) Unless the github action mangled this, I'm not seeing the "fixed issue with having edited fields be off screen" fix happening... if I press on an edit box, and it's not basically at the top of the screen, it scrolls 'up' out of sight. Otherwise, on the TX16s everything else seems good though - keyboard is centred, get a highlight when pressing a key (although naturally your finger is on it, so probably works best with a stylus), backspace looks ok and 'wrong keypress' seems to be a nonissue.

pfeerick commented:
NV14 build seems to be the same

JimB40 commented:
Guys. This touches UI.
1. Is it kind of standard libui keyboard
2.How can I check this feature? Artifact?

pfeerick commented:
Yes - "standard" libopenui keyboard - the one that has been visible on the NV14, and yes available via Checks tab.

kevinkoenig commented:
> First impression... this is really good! :) Unless the github action mangled this, I'm not seeing the "fixed issue with having edited fields be off screen" fix happening... if I press on an edit box, and it's not basically at the top of the screen, it scrolls 'up' out of sight. Otherwise, on the TX16s everything else seems good though - keyboard is centred, get a highlight when pressing a key (although naturally your finger is on it, so probably works best with a stylus), backspace looks ok and 'wrong keypress' seems to be a nonissue.

can you explain a little better how you got a field to scroll off screen.  I tried to make that never happen.  I will go back in and take a look

pfeerick commented:
I basically just picked a field like model name or timer name by touch, and it promptly scrolled out of view when the keyboard showed. 

Maybe if you could try a build from the checks tab - as if that is behaving differently to a local build of your repo then it's not properly building your changes...

kevinkoenig commented:
> I basically just picked a field like model name or timer name by touch, and it promptly scrolled out of view when the keyboard showed.
> 
> Maybe if you could try a build from the checks tab - as if that is behaving differently to a local build of your repo then it's not properly building your changes...

i know this is obvious but all of the magic is in libopenui.  which is a separate pull request.  when I run it on my tx16s and simu it does not exhibit that behavior.

kevinkoenig commented:
i rebased the code base as suggested above.  also by default the linked library libopenui defaults to the wrong branch.  to properly test this you need to change the remote URL first and set the branch.  i don't know how to make this happen automatically.

pfeerick commented:
I'll  reply here so it's visible (rather than on Discord)... I think you spotted the problem... The GitHub actions build will be look at your submodule, but not your branch, so it will need to be self-built for now. I'll do that for TX16S and NV14 and share the bins in a few hours when I get home from flying ;)

kevinkoenig commented:
Great thanks

Sent from my iPhone

> On Aug 28, 2021, at 5:18 PM, Peter Feerick ***@***.***> wrote:
> 
> ﻿
> I'll reply here so it's visible (rather than on Discord)... I think you spotted the problem... The GitHub actions build will be look at your submodule, but not your branch, so it will need to be self-built for now. I'll do that for TX16S and NV14 and share the bins in a few hours when I get home from flying ;)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

pfeerick commented:
I think to make this auto-build, it may just be a matter of changing `master` in `.gitmodules` on line 4 to also point to your software-keyboard branch on libopenui (which will also need to be switched back on merge). 

Here are zipped binaries for the TX16S and for the NV14 for the most recent commit. I've tried the TX16s version, and edit fields are now staying in view, and the keyboard is a lot less error prone, so it looks like I was using the wrong version before! 😁 

[tx16s_software-keyboard_0ea76b12.zip](https://github.com/EdgeTX/edgetx/files/7071832/tx16s_software-keyboard_0ea76b12.zip)

[nv14_software-keyboard_0ea76b12.zip](https://github.com/EdgeTX/edgetx/files/7071831/nv14_software-keyboard_0ea76b12.zip)

kevinkoenig commented:
> I think to make this auto-build properly, it may just be a matter of changing `master` in `.gitmodules` on line 4 to also point to the `software-keyboard` branch on your fork of libopenui (which will also need to be switched back on merge) ;)
> 
> Here are zipped binaries for the TX16S and for the NV14 for the most recent commit. I've tried the TX16s version, and edit fields are now staying in view, and the keyboard is a lot less error prone, so it looks like I was using the wrong version before! 😁
> 
> [tx16s_software-keyboard_0ea76b12.zip](https://github.com/EdgeTX/edgetx/files/7071832/tx16s_software-keyboard_0ea76b12.zip)
> 
> [nv14_software-keyboard_0ea76b12.zip](https://github.com/EdgeTX/edgetx/files/7071831/nv14_software-keyboard_0ea76b12.zip)

I updated gitmodules as you suggested.  And I rebased it yesterday.  let me know what else I need to do to complete this PR

rotorman commented:
Still seems not to be rebased on the main branch head unfortunately.

rotorman commented:
Gave this PR a whirl - looking really nice. Well done!

Here a demo while editing model name with virtual touch in simu:

![EdgeTX_virtual_keyboard](https://user-images.githubusercontent.com/21011587/131257393-bd3d65dd-adc6-4d98-b276-370f985e7926.png)

raphaelcoeffic commented:
@kevinkoenig In case you need my help for the git-tricks, there are basically 2 options: either you give me "write access" to your repo, or I port your commits into a new PR hosted here directly. Your choice ;-)

raphaelcoeffic commented:
> @kevinkoenig In case you need my help for the git-tricks, there are basically 2 options: either you give me "write access" to your repo, or I port your commits into a new PR hosted here directly. Your choice ;-)


After some checking, it seems it's almost perfect: I merged your libopenui PR, now all you need to do is to change the submodule pointer back to what is was.

So basically in `.gitmodules`, change from this:
```
[submodule "libopenui"]
	path = radio/src/thirdparty/libopenui
	url = https://github.com/kevinkoenig/libopenui.git
	branch = software-keyboard
```

Back to this:
```
[submodule "libopenui"]
	path = radio/src/thirdparty/libopenui
	url = https://github.com/EdgeTX/libopenui.git
	branch = master
```

kevinkoenig commented:
Will do when I get back to house.

Sent from my iPhone

> On Aug 29, 2021, at 12:38 PM, Raphael Coeffic ***@***.***> wrote:
> 
> ﻿
> @kevinkoenig In case you need my help for the git-tricks, there are basically 2 options: either you give me "write access" to your repo, or I port your commits into a new PR hosted here directly. Your choice ;-)
> 
> After some checking, it seems it's almost perfect: I merged your libopenui PR, now all you need to do is to change the submodule pointer back to what is was.
> 
> So basically in .gitmodules, change from this:
> 
> [submodule "libopenui"]
> 	path = radio/src/thirdparty/libopenui
> 	url = https://github.com/kevinkoenig/libopenui.git
> 	branch = software-keyboard
> Back to this:
> 
> [submodule "libopenui"]
> 	path = radio/src/thirdparty/libopenui
> 	url = https://github.com/EdgeTX/libopenui.git
> 	branch = master
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

kevinkoenig commented:
> @kevinkoenig In case you need my help for the git-tricks, there are basically 2 options: either you give me "write access" to your repo, or I port your commits into a new PR hosted here directly. Your choice ;-)

I think I rebased correctly but i am still not sure.  I have invited you to the repository.  I would greatly appreciate the help in getting this PR completed.

pfeerick commented:
@raphaelcoeffic I've edited this PR so that only the needed changes are left, but the submodule is being stubborn and not allowing it to merge... I was going to try and update the submodule hash the repo is using, but that will automatically break as the changes to the softkeyboard also need this PRs compile flags to work/compile, so I'm going to admit I'm too chicken to go further, especially since I don't know what I'm doing by this point! 😆

rotorman commented:
I am thinking out loud if purely reverting here the radio/src/thirdparty/libopenui would already get us to what we want - the libopenui patches themselves have already been merged in libopenui repo as far as I can see?

pfeerick commented:
I can certainly try... at least you're thinking what I was thinking but didn't want to fiddle any more with it blindly... In this case, I will try pushing the same submodule commit as the repo is currently using... as the current state is to bump it to the current submodule version.

pfeerick commented:
@rotorman You're a genuis! Looks like PRs can't bump the repo submodule version.

rotorman commented:
Haah, couldn't be further away from genius - more beginners luck I would say. I am still learning Git myself...
